06_puppy/mohankrishna/types.go

```diff
@@ -0,0 +1,31 @@
+package main
+
+// Error codes
+const (
+	ErrInvalidInput int = iota
+	ErrDuplicate
+	ErrNotFound
+)
+
+type Error struct {
+	Code    int
+	Message string
+}
+
+func (e *Error) Error() string {
+	return e.Message
+}
+
+type Puppy struct {
+	ID     uint32
+	Breed  string
+	Colour string
+	Value  int
+}
+
+type Storer interface {
+	CreatePuppy(puppy *Puppy) error
+	ReadPuppy(ID uint32) (*Puppy, error)
+	UpdatePuppy(ID uint32, puppy *Puppy) error
+	DeletePuppy(ID uint32) (bool, error)
```

Comment:
It is questionable as to whether `DeletePuppy` should return two values. The way it is currently being used means there is no overlap between the bool and the error, so you may as well just return an error. If the error is `nil`, then all is ok. If it is not, then it's not ok.

This might change if you decide that it is not an error to delete a puppy that does not exist. Then `ok` returns something valuable. The error would then be for implementations where trying to delete leads to errors outside your control, such as when using an external database. If the connection to the database is lost or some other database error, an error would be returned. This would mean your current implementations have no cases where it would return an error.